active/0051-strategies.md

```diff
@@ -0,0 +1,368 @@
+
+Strategies get access to the same input tree, they are not chained off of each
+other and they don't run in parallel.
+
+### Concatenation Strategy
+
+It is used by Ember CLI privately in the latest `canary` version to produce
+`assets/app-name.js` and `assets/vendor.js`. It takes a broccoli tree and
+returns a new concatenated broccoli tree.
+
+```javascript
+const concat = require('broccoli-concat');
+
+class ConcatenationStrategy {
+  constructor(options) { this.options = options; }
+
+  toTree(assembler, inputTree) {
+    return concat(inputTree, this.options);
+  }
+}
+```
+
+Where `this.options` mimics [`broccoli-concat`
+options](https://github.com/broccolijs/broccoli-concat#advanced-usage).
+
+### Example of possible strategies
+
+Below are some of the strategies that could be implemented.
+
+#### Assets Split Strategy
+
+One of the techniques for improving site speed is isolating changes throughout
+application deployments. Assuming  the application assets are uploaded to CDN,
+the reasoning is very simple: if `ember.js` or `jQuery` (possibly along with
+other third party libraries) don't change with every deployment, why bust CDN
+cache for them?
+
+This strategy could be provided out of the box by Ember CLI as a part of site
+speed improvement story.
+
+```javascript
+class AssetsSplitStrategy {
+  constructor(options) {
+    this.options = options;
+  }
+
+  toTree(assembler, inputTree) {
+    return concat(inputTree, {
+      headerFiles: [
+        jQueryPath, emberPath
+      ],
+      outputFile: 'vendor-static.js'
+    });
+  }
+}
+```
+
+#### Debug/Analyse Strategy
+
+Yet another application of strategies would be to run different analysis on the
+Ember applications. Take
+[broccoli-concat-analyser](https://github.com/stefanpenner/broccoli-concat-analyser),
+for example. This could be a strategy and easily encorporated into the build.
+
+#### Engine Strategy
+
+Introducing `Strategy` as a concept allows us to re-think the way we approach
+Ember Engines as well. We can start thinking about engines as a strategy of
+certain type. In fact, it's very similar to `ConcatenationStrategy` described
+above.
+
++ reduces the API surface (as opposed to `ember-engines` and
+  `ember-asset-loader`, only one strategy)
++ makes things explicit, meaning that you have to register a strategy
+
+## Registering custom strategies
+
+Ember applications would be able to register custom strategies via
+`ember-cli-build.js` file by passing a `strategies` property to the constructor.
+
+Passing `strategies` as an option to the constructor will clobber default
+strategies that Ember CLI provides out of the box (they mimic existing behaviour
+of creating `dist` folder with final assets).
+
+However, Ember CLI will expose an array of default strategies that people can
+take advantage of. Note, that `defaultStrategies` will be freezed so you can't
+push directly onto it. One would need to create a new array.
```

Comment:
I'm assuming the processing of these is simply in they order they appear in the array, though I was curious if there would be need for some sort of priority system to allow some finer control over order of operations in case it was needed.

Reply:
I would assume that priority system would be implemented in the strategies themselves and would be managed by the assembler when processing the strategies (sort of like how addons are handled currently)?

Reply:
I mention this in "Unresolved Question" section. There might be an ordering problem but before coming up with a solution, I think we should surface a real use case first. Add-ons don't have a priority system and things "just work".

> I would assume that priority system would be implemented in the strategies themselves 

That's one way to implement it, I would stay clear for a particular implementation until we know more about constraints.

Reply:
Actually, in the addon config (inside the `package.json`), you can define an array of `before` and `after` to ensure that addons are resolved in the desired order, which then is managed by the dependency graph:
https://github.com/ember-cli/ember-cli/blob/master/lib/models/addons-factory.js#L41

We use this strategy for `ember-cli-legacy-blueprints`, to ensure that if `ember-core` is installed, its blueprints will win:
https://github.com/ember-cli/ember-cli-legacy-blueprints/blob/master/package.json#L55